trusted-services.mk

```diff
@@ -0,0 +1,130 @@
+################################################################################
+# Paths to Trusted Services source and output
+################################################################################
+TS_PATH			?= $(ROOT)/trusted-services
+TS_BUILD_PATH		?= $(OUT_PATH)/ts-build
+TS_INSTALL_PREFIX	?= $(OUT_PATH)/ts-install
+
+################################################################################
+# Secure Partitions
+################################################################################
+.PHONY: ffa-sp-all ffa-sp-all-clean ffa-sp-all-realclean
+
+optee-os-common: ffa-sp-all
+optee-os-clean: ffa-sp-all-clean
+
+ffa-sp-all-realclean:
+	rm -rf $(TS_INSTALL_PREFIX)/opteesp
+
+# Helper macro to build and install Trusted Services Secure Partitions (SPs).
+# Invokes CMake to configure, and make to build and install the SP. (CMake's
+# Makefile generator backend is used, we can run make in the build directory).
+# Adds the SP output image to the optee_os_sp_paths list and complies the SP
+# manifest dts to dtb.
+#
+# For information about the additional dependencies of the project, please see
+# https://trusted-services.readthedocs.io/en/latest/developer/software-requirements.html
+#
+# Parameter list:
+# 1 - SP deployment name (e.g. internal-trusted-storage, crypto, etc.)
+# 2 - SP canonical UUID (e.g. dc1eef48-b17a-4ccf-ac8b-dfcff7711b14)
+# 3 - SP additional build flags (e.g. -DTS_PLATFORM=<...>)
+define build-sp
+.PHONY: ffa-$1-sp
+ffa-$1-sp:
+	CROSS_COMPILE=$(AARCH64_CROSS_COMPILE) cmake -G"Unix Makefiles" \
```

Comment:
Would it make sense to use `CROSS_COMPILE_S_USER` here, instead of `AARCH64_CROSS_COMPILE`?
On one hand it would look cleaner, but on the other hand the SPs only work on AArch64 so probably it's better to explicitly show that here.

Reply:
Yes, `CROSS_COMPILE_S_USER` is the proper variable to use here. If 32-bit isn't supported then just add:
```
ifneq ($(COMPILE_S_USER),64)
$(error SPs can only be built for Aarch64)
endif
```
(just before `define build-sp`)

Reply:
> `CROSS_COMPILE_S_USER` contains ccache at the beginning, and Trusted Services only expects the compiler path prefix in this parameter.

...which is a bug, arguably :wink:

> So I had to add a `$(subst ...)` too, but I think it's not too bad.

I'd prefer the TS build to be fixed but it's not a big deal.

Reply:
The cmake idiom is to manage ccache configuration trough `CMAKE_C_COMPILER_LAUNCHER`.
Although we could fix the compiler check in TS cmake scripts, but if someone sets `CMAKE_C_COMPILER_LAUNCHER` and both has ccache in `CROSS_COMPILE`, that won't work. Similarly cmake fails if `CC` environment variable has ccache.
We thought aligning to cmake idioms is better, even if the implementation looks in the makefile less nice.

Reply:
@gyuri-szing alright, thanks for the explanation.